### ISSUE TITLE
TCP Steal/mirror doesn't work with shared sockets

### ISSUE THREAD

**Issue by aviramha** [metalbear-co/mirrord#864]
### Bug Description

When a parent process binds a socket, then children sockets listen/accept on that socket mirrord doesn't intercept those calls because the fd isn't managed (doesn't exist in our fork).
Using this issue to document use cases where this happens for now:

- uvicorn when using `--reload` flag.
- hypercorn
		

### Steps to Reproduce

```python
from fastapi import FastAPI
import sys
app = FastAPI()


@app.get("/")
async def root():
    print("cake")
    return {"message": "Hello World"}


@app.get("/hello/{name}")
async def say_hello(name: str):
    print("pake")
    return {"message": f"Hello {name}"}

```

run with `mirrord exec --target pod/pod uvicorn --  --reload --port 80 main:app`

### Your operating system and version

macOS

### Local process

python

**Comment by aviramha** [comment]
Possible solutions:

1. hooking fork/exec, then passing down the `SOCKET/FILES` fd struct as environment variable for fds that aren't CLOEXEC
2. Use shared memory (shmem) for `SOCKET/FILES` (dunno about this..)
3. on `listen` with missing fd (this happens on uvicorn + reloader case) use getsockname to resolve addr and then decide if to add the socket and backfill the information. This is pretty good approach but has some caveats:
  - It would work only on the listen case, not if listen is called in parent and accept in child (not sure if this happens and what happens in this case)
  - the socket information we would get from getsockname would be the "fake port" and not the one the remote would need to listen for, so maybe this is impossible ;(

**Comment by aviramha** [comment]
Please let us know if this is impacting anyone so we can know if we need to prioritize it.

**Comment by danielloader** [comment]
This would probably affect me if not already blocked by this [issue](https://github.com/metalbear-co/mirrord/issues/1618)

**Comment by yoav-orca** [comment]
I'm not blocked by this, but it does affect my productivity.
MacOS M1 Pro

**Comment by aviramha** [comment]
> I'm not blocked by this, but it does affect my productivity.
> 
> MacOS M1 Pro

Hey, does it still happen to you on latest version?

**Comment by yoav-orca** [comment]
I'm running `3.75.0` and I'm not getting traffic with `--reload` and I do get traffic without it

**Comment by aviramha** [comment]
Also doesn't work when using Uvicorn with more than one worker process.

**Comment by surbas** [comment]
Happening to me on 3.84.1. I am not using the operator. Not sure if that matters.

**Comment by brandfocus** [comment]
Ran into the same issue. Might be worth flagging this in the docs somewhere until its fixed

**Comment by meowjesty** [comment]
I've been exploring this, and here are a few solutions I can think of:

- :earth_americas:  how can it be global, if the code is flat

The main issue we have is that our `sockfd`s are stored in a per-layer global `SOCKETS`, which are not being shared between forks. So the proposal here is to just move the `SOCKETS` global to the `intproxy`, and change every operation that calls `SOCKETS.[get|take](&sockfd)` to use `make_proxy_request_with_response` (or something similar).

A problem with this solution is that now every `sockfd` becomes shared between `exec`s and `fork`s, which is not ideal. Another (minor) problem is that we must manually remember to call `insert` after `take`, as some operations call `SOCKETS.take` to modify the socket, meh.

- :donkey: every masterpiece has its cheap copy

Instead of moving the whole `SOCKETS` to `intproxy`, whenever we detect a `fork` (new layer connection?), we send a copy of the `SOCKETS` minus the ones with `CLOEXEC` to the `intproxy`, which in turn sets the new layer connection's `SOCKETS` to this curated copy.

- :peach: thicc intproxy

Delegate more work to the `intproxy`, by moving the `socket/ops` there, and creating some sort of `MirrordSocket` API that keeps an association between a `sockfd` and the `pid` that created it, so we can avoid leaking `sockfd`s.

That's it, so far. I'm leaning more towards the :donkey: solution, as :earth_americas: is a bug waiting to happen, and :peach: is too big of a pointless refactor.

**Comment by aviramha** [comment]
🫏 is what would make most sense and is how it works in terms of OS level too.
We can actually pass all `SOCKETS` downstream and iterate each fd and see if it's still open (so we'll just follow OS logic if cloexec or not)

**Comment by aviramha** [comment]
Fixed in #2610

**Comment by meowjesty** [comment]
Looks like the problem is still there on macOS + vscode.
